nixos/modules/services/continuous-integration/gitlab-runner.nix

```diff
@@ -1,160 +1,423 @@
-{ config, lib, pkgs, ... }:
-
+{ config, pkgs, lib, ... }:
 with lib;
-
 let
   cfg = config.services.gitlab-runner;
-  configFile =
-    if (cfg.configFile == null) then
-      (pkgs.runCommand "config.toml" {
-        buildInputs = [ pkgs.remarshal ];
-        preferLocalBuild = true;
-      } ''
-        remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
-          > $out
-      '')
-    else
-      cfg.configFile;
   hasDocker = config.virtualisation.docker.enable;
+  hashedServices = with builtins; (mapAttrs' (name: service: nameValuePair
+    "${name}_${config.networking.hostName}_${unsafeDiscardStringContext (
+        substring 0 12
+        (hashString "md5" (toJSON service)))}"
+      service)
+    cfg.services);
+  configPath = "/var/lib/gitlab-runner/.gitlab-runner/config.toml";
+  configureScript = pkgs.writeShellScriptBin "gitlab-runner-configure" (
+    if (cfg.configFile != null) then "cp ${cfg.configFile} ${configPath}" else ''
+    # remove no longer existing services
+    gitlab-runner verify --delete
+
+    # current and desired state
+    NEEDED_SERVICES=$(echo ${concatStringsSep " " (attrNames hashedServices)} | tr " " "\n")
+    REGISTERED_SERVICES=$(gitlab-runner list 2>&1 | grep 'Executor' | awk '{ print $1 }')
+
+    # difference between current and desired state
+    NEW_SERVICES=$(grep -vxF -f <(echo "$REGISTERED_SERVICES") <(echo "$NEEDED_SERVICES") || true)
+    OLD_SERVICES=$(grep -vxF -f <(echo "$NEEDED_SERVICES") <(echo "$REGISTERED_SERVICES") || true)
+
+    # register new services
+    ${concatStringsSep "\n" (mapAttrsToList (name: service: ''
```

Comment:
we should have some assertions only arguments valid for the given executor are passed - I doubt specifying `--docker-image` for the `shell` executor makes sense, it might just throw an error.

Reply:
Done.